iio/iio_demo/iio_demo.c

```diff
@@ -0,0 +1,416 @@
+{
+	struct iio_demo_device *demo_device;
+	uint32_t index, addr;
+	uint16_t *buf16;
+
+	if (!iio_inst)
+		return FAILURE;
+
+	if (!buf)
+		return FAILURE;
+
+	buf16 = (uint16_t *)buf;
+	demo_device = (struct iio_demo_device *)iio_inst;
+	addr = demo_device->ddr_base_addr + offset;
+	for(index = 0; index < bytes_count; index += 2) {
+		uint32_t *local_addr = (uint32_t *)(addr + index * 2);
+		*local_addr = (buf16[index + 1] << 16) | buf16[index];
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Read chunk of data from RAM to pbuf.
+ * Call "iio_demo_transfer_dev_to_mem" first.
+ * This function is probably called multiple times by libtinyiiod after a
+ * "iio_demo_transfer_dev_to_mem" call, since we can only read "bytes_count"
+ * bytes at a time.
+ * @param iio_inst - Physical instance of a device.
+ * @param pbuf - Buffer where value is stored.
+ * @param offset - Offset to the remaining data after reading n chunks.
+ * @param bytes_count - Number of bytes to read.
+ * @param ch_mask - Opened channels mask.
+ * @return bytes_count or negative value in case of error.
+ */
+static ssize_t iio_demo_read_dev(void *iio_inst, char *pbuf, size_t offset,
+				 size_t bytes_count, uint32_t ch_mask)
+{
+	struct iio_demo_device *demo_device;
+	uint32_t i, j = 0, current_ch = 0;
+	uint16_t *pbuf16;
+	size_t samples;
+
+	if (!iio_inst)
+		return FAILURE;
+
+	if (!pbuf)
+		return FAILURE;
+
+	demo_device = (struct iio_demo_device *)iio_inst;
+	pbuf16 = (uint16_t*)pbuf;
+	samples = (bytes_count * demo_device->num_channels) / hweight8(
+			  ch_mask);
+	samples /= 2; /* because of uint16_t *pbuf16 = (uint16_t*)pbuf; */
+	offset = (offset * demo_device->num_channels) / hweight8(ch_mask);
+
+	for (i = 0; i < samples; i++) {
+
+		if (ch_mask & BIT(current_ch)) {
+			pbuf16[j] = *(uint16_t*)(demo_device->ddr_base_addr + offset + i * 2);
+			j++;
+		}
+
+		if (current_ch + 1 < demo_device->num_channels)
+			current_ch++;
+		else
+			current_ch = 0;
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Get xml corresponding to device.
+ * @param xml - Xml containing description of a device.
+ * @param iio_dev - Structure describing a device, channels and attributes.
+ * @return SUCCESS in case of success or negative value otherwise.
+ */
+static ssize_t iio_demo_get_xml(char** xml, struct iio_device *iio_dev)
+{
+
+	*xml = calloc(1, strlen(demo_xml) + 1);
+	if (!(*xml))
+		return FAILURE;
+
+	memcpy(*xml, demo_xml, strlen(demo_xml));
+
+	return SUCCESS;
+}
+
+/**
+ * @brief iio demo init function, registers a demo .
+ * @param desc - Descriptor.
+ * @param init - Configuration structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t iio_demo_init(struct iio_demo_desc **desc,
+		      struct iio_demo_init_param *init)
+{
+	struct iio_interface *iio_interface;
+	struct iio_device * iio_device;
+	int32_t status;
+	struct iio_demo_device *iio_demo_device_inst = (struct iio_demo_device *)calloc(
+				1,
+				sizeof(struct iio_demo_device));
```

Comment:
According with linux coding style we should use sizeof in calloc this way:
`calloc(1, sizeof(*iio_demo_device_inst))`
same for all